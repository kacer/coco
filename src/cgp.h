/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include <stdbool.h>
#include <stdio.h>

#include "cgp_config.h"


#define CGP_FUNC_INPUTS 2
#define CGP_NODES (CGP_COLS * CGP_ROWS)
#define CGP_CHR_OUTPUTS_INDEX ((CGP_FUNC_INPUTS + 1) * CGP_NODES)
#define CGP_CHR_LENGTH (CGP_CHR_OUTPUTS_INDEX + CGP_OUTPUTS)

typedef unsigned char cgp_value_t;
typedef double cgp_fitness_t;

#define CGP_FUNC_COUNT 16
typedef enum
{
    c255 = 0,   // 255
    identity,   // a
    inversion,  // 255 - a
    b_or,       // a or b
    b_not1or2,  // (not a) or b
    b_and,      // a and b
    b_nand,     // not (a and b)
    b_xor,      // a xor b
    rshift1,    // a >> 1
    rshift2,    // a >> 2
    swap,       // a <-> b
    add,        // a + b
    add_sat,    // a +S b
    avg,        // (a + b) >> 1
    max,        // max(a, b)
    min,        // min(a, b)
} _cgp_func;


static inline char* cgp_func_name(_cgp_func f) {
    char *func_names[] = {
        " FF ",     // 255
        "  a ",     // a
        "FF-a",     // 255 - a
        " or ",     // a or b
        "~1|2",     // (not a) or b
        " and",     // a and b
        "nand",     // not (a and b)
        " xor",     // a xor b
        "a>>1",     // a >> 1
        "a>>2",     // a >> 2
        "swap",     // a <-> b
        " +  ",     // a + b
        " +S ",     // a +S b
        " avg",     // (a + b) >> 1
        " max",     // max(a, b)
        " min",     // min(a, b)
    };
    return func_names[f];
}


/**
 * One CGP node (function block)
 */
typedef struct {
    int inputs[CGP_FUNC_INPUTS];
    _cgp_func function;
} _cgp_node;


/**
 * Chromosome
 */
typedef struct {
    bool has_fitness;
    cgp_fitness_t fitness;
    _cgp_node nodes[CGP_COLS * CGP_ROWS];
    int outputs[CGP_OUTPUTS];
} _cgp_chr;
typedef _cgp_chr* cgp_chr;


/**
 * Population
 */
typedef struct {
    int size;
    int generation;
    cgp_fitness_t best_fitness;
    int best_chr_index;
    cgp_chr *chromosomes;
} _cgp_pop;
typedef _cgp_pop* cgp_pop;


/**
 * Dump formats
 */
typedef enum {
    asciiart,
    compat,
    readable,
} cgp_dump_format;


/**
 * CGP problem type
 */
typedef enum {
    minimize,
    maximize,
} cgp_problem_type;


/**
 * Fitness function
 */
typedef cgp_fitness_t (*cgp_fitness_func)(cgp_chr chromosome);


/**
 * Initialize CGP internals
 * @param Fitness function to use
 * @param Type of solved problem
 */
void cgp_init(cgp_fitness_func fitness, cgp_problem_type type);


/**
 * Deinitialize CGP internals
 */
void cgp_deinit();


/**
 * Create a new chromosome
 * @return
 */
cgp_chr cgp_create_chr();


/**
 * Clear memory associated with given chromosome
 * @param chr
 */
void cgp_destroy_chr(cgp_chr chr);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_replace_chr(cgp_chr chr, cgp_chr replacement);


/**
 * Mutate given chromosome
 * @param chr
 * @param max_changed_genes
 */
void cgp_mutate_chr(cgp_chr chr, int max_changed_genes);


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 */
void cgp_get_output(cgp_chr chr, cgp_value_t *inputs, cgp_value_t *outputs);


/**
 * Calculate fitness of given chromosome, but only if its `has_fitness`
 * attribute is set to `false`
 * @param chr
 */
cgp_fitness_t cgp_evaluate_chr(cgp_chr chr);


/**
 * Calculate fitness of given chromosome, regardless of its `has_fitness`
 * value
 * @param chr
 */
cgp_fitness_t cgp_reevaluate_chr(cgp_chr chr);


/**
 * Create a new CGP population with given size
 * @param  size
 * @return
 */
cgp_pop cgp_create_pop(int size);


/**
 * Clear memory associated with given population (including its chromosomes)
 * @param pop
 */
void cgp_destroy_pop(cgp_pop pop);


/**
 * Calculate fitness of whole population, using `cgp_evaluate_chr`
 * @param chr
 */
static inline void cgp_evaluate_pop(cgp_pop pop);


/**
 * Calculate fitness of whole population, using `cgp_evaluate_chr`
 * in single thread
 * @param chr
 */
void _cgp_evaluate_pop_simple(cgp_pop pop);


/**
 * Calculate fitness of whole population, using `cgp_evaluate_chr`
 * using one thread per chromosome
 * @param chr
 */
void _cgp_evaluate_pop_pthread(cgp_pop pop);


#ifdef CGP_USE_PTHREAD

    static inline void cgp_evaluate_pop(cgp_pop pop) {
        _cgp_evaluate_pop_pthread(pop);
    }

#else

    static inline void cgp_evaluate_pop(cgp_pop pop) {
        _cgp_evaluate_pop_simple(pop);
    }

#endif


/**
 * Advance population to next generation
 * @param pop
 * @param mutation_rate
 */
void cgp_next_generation(cgp_pop pop, int mutation_rate);


/**
 * Dumps chromosome to given file pointer in given format
 * @param fp
 */
void cgp_dump_chr(cgp_chr chr, FILE *fp, cgp_dump_format fmt);


/**
 * Dumps chromosome to given file in CGP-viewer compatible format
 * @param chr
 * @param fp
 */
void cgp_dump_chr_compat(cgp_chr chr, FILE *fp);


/**
 * Dumps chromosome to given file in more human-friendly fashion
 * @param chr
 * @param fp
 */
void cgp_dump_chr_readable(cgp_chr chr, FILE *fp);


/**
 * Dumps chromosome to given file as an ASCII-art, which looks like this:
 *
 *      .------------------------------------.
 *      |      .----.            .----.      |
 * [ 0]>| [ 1]>|    |>[ 4]  [ 0]>|    |>[ 6] |>[ 6]
 * [ 1]>| [ 3]>|  a |       [ 4]>| xor|      |
 * [ 2]>|      '----'            '----'      |
 * [ 3]>|      .----.            .----.      |
 *      | [ 3]>|    |>[ 5]  [ 4]>|    |>[ 7] |
 *      | [ 0]>|a>>1|       [ 5]>|nand|      |
 *      |      '----'            '----'      |
 *      '------------------------------------'
 *
 * @param chr
 * @param fp
*/
void cgp_dump_chr_asciiart(cgp_chr chr, FILE *fp);


static inline int cgp_dump_chr_asciiart_width() {
    return 5 + CGP_COLS * 18 + 5;
}


static inline int cgp_dump_chr_asciiart_height() {
    return 1 + CGP_ROWS * 4 + 1;
}


/**
 * Loads chromosome from given file stored in CGP-viewer compatible format
 * @param chr
 * @param fp
 * @return 0 on success, -1 on file format error, -2 on incompatible CGP config
 */
int cgp_load_chr_compat(cgp_chr chr, FILE *fp);


/**
 * Dumps whole population to given file pointer with chromosomes in
 * CGP-viewer compatible format
 * @param pop
 * @param fp
 */
void cgp_dump_pop_compat(cgp_pop pop, FILE *fp);


/**
 * Loads whole population from given file with chromosomes stored in
 * CGP-viewer compatible format
 * @param pop
 * @param fp
 * @return 0 on success, -1 on file format error, -2 on incompatible CGP config
 */
int cgp_load_pop_compat(cgp_pop *pop_ptr, FILE *fp);


/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_node_index(int col, int row)
{
    return CGP_ROWS * col + row;
}


/**
 * Returns node column based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_col(int index)
{
    return index / CGP_ROWS;
}


/**
 * Returns node row based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_row(int index)
{
    return index % CGP_ROWS;
}
