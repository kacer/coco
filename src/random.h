/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <time.h>


static inline void rand_init()
{
    srand(time(0));
}

static inline void rand_init_seed(unsigned int seed)
{
    srand(seed);
}

static inline unsigned int rnd()
{
    return rand();
}


/**
 * Generates random number between low and high, inclusive
 * @param  low
 * @param  high
 * @return
 */
static inline int rand_range(int low, int high)
{
    return rand() % (high - low + 1) + low;
}


/**
 * Generates random number between low and high, inclusive
 * @param  low
 * @param  high
 * @return
 */
static inline unsigned int rand_urange(unsigned int low, unsigned int high)
{
    return rand() % (high - low + 1) + low;
}


/**
 * Returns randomly chosen number from the list of signed integers
 * @param  length
 * @param  choices
 * @return
 */
static inline int rand_schoice(int length, int choices[])
{
    unsigned int index = rand_range(0, length - 1);
    return choices[index];
}



/**
 * Returns randomly chosen number from the list of unsigned integers
 * @param  length
 * @param  choices
 * @return
 */
static inline unsigned int rand_uchoice(unsigned int length, unsigned int choices[])
{
    unsigned int index = rand_range(0, length - 1);
    return choices[index];
}
