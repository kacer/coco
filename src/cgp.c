/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#include <stdlib.h>
#include <string.h>

#include "cgp.h"
#include "random.h"


void cgp_randomize_gene(cgp_chr chr, int gene);


typedef struct {
    int size;
    int *values;
} int_array;

int_array allowed_vals[CGP_COLS];
cgp_fitness_func fitness_function;
cgp_problem_type problem_type;


#ifdef TEST_RANDOMIZE
    #define TEST_RANDOMIZE_PRINTF(...) printf(__VA_ARGS__)
#else
    #define TEST_RANDOMIZE_PRINTF(...) /* printf(__VA_ARGS__) */
#endif


/**
 * Returns true if WHAT is better or same as COMPARED_TO
 * @param  what
 * @param  compared_to
 * @return
 */
static inline bool cgp_is_better_or_same(cgp_fitness_t what, cgp_fitness_t compared_to) {
    return (problem_type == minimize)? what <= compared_to : what >= compared_to;
}



/**
 * Initialize CGP internals
 * @param Fitness function to use
 * @param Type of solved problem
 */
void cgp_init(cgp_fitness_func fitness, cgp_problem_type type)
{
    // remember fitness function and problem type
    fitness_function = fitness;
    problem_type = type;

    // calculate allowed values of node inputs in each column
    for (int x = 0; x < CGP_COLS; x++) {

        // range of outputs which can be connected to node in i-th column
        // maximum is actually by 1 larger than maximum allowed value
        // (so there is `val < maximum` in the for loop below instead of `<=`)
        int minimum = CGP_ROWS * (x - CGP_LBACK) + CGP_INPUTS;
        if (minimum < CGP_INPUTS) minimum = CGP_INPUTS;
        int maximum = CGP_ROWS * x + CGP_INPUTS;

        int size = CGP_INPUTS + maximum - minimum;
        allowed_vals[x].size = size;
        allowed_vals[x].values = (int*) malloc(sizeof(int) * size);

        int key = 0;
        // primary input indexes
        for (int val = 0; val < CGP_INPUTS; val++, key++) {
            allowed_vals[x].values[key] = val;
        }

        // nodes to the left output indexes
        for (int val = minimum; val < maximum; val++, key++) {
            allowed_vals[x].values[key] = val;
        }
    }


#ifdef TEST_INIT
    for (int x = 0; x < CGP_COLS; x++) {
        printf ("x = %d: ", x);
        for (int y = 0; y < allowed_vals[x].size; y++) {
            if (y > 0) printf(", ");
            printf ("%d", allowed_vals[x].values[y]);
        }
        printf("\n");
    }
#endif
}


/**
 * Deinitialize CGP internals
 */
void cgp_deinit()
{
    for (int x = 0; x < CGP_COLS; x++) {
        free(allowed_vals[x].values);
    }
}


/**
 * Create a new chromosome
 * @return
 */
cgp_chr cgp_create_chr()
{
    cgp_chr new_chr = (cgp_chr) malloc(sizeof(_cgp_chr));
    new_chr->has_fitness = false;

    for (int i = 0; i < CGP_CHR_LENGTH; i++) {
        cgp_randomize_gene(new_chr, i);
    }

    return new_chr;
}


/**
 * Clear memory associated with given chromosome
 * @param chr
 */
void cgp_destroy_chr(cgp_chr chr)
{
    free(chr);
}


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_replace_chr(cgp_chr chr, cgp_chr replacement)
{
    memcpy(chr->nodes, replacement->nodes, sizeof(_cgp_node) * CGP_NODES);
    memcpy(chr->outputs, replacement->outputs, sizeof(int) * CGP_OUTPUTS);
    chr->fitness = replacement->fitness;
    chr->has_fitness = replacement->has_fitness;
}


/**
 * Mutate given chromosome
 * @param chr
 * @param max_changed_genes
 */
void cgp_mutate_chr(cgp_chr chr, int max_changed_genes)
{
    int genes_to_change = rand_range(0, max_changed_genes);
    for (int i = 0; i < genes_to_change; i++) {
        int gene = rand_range(0, CGP_CHR_LENGTH - 1);
        cgp_randomize_gene(chr, gene);
    }
    chr->has_fitness = false;
}


#define SWAP(A, B) (((A & 0x0F) << 4) | ((B & 0x0F)))
#define ADD_SAT(A, B) ((A > 0xFF - B) ? 0xFF : A + B)
#define MAX(A, B) ((A > B) ? A : B)
#define MIN(A, B) ((A < B) ? A : B)


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 */
void cgp_get_output(cgp_chr chr, cgp_value_t *inputs, cgp_value_t *outputs)
{
    cgp_value_t inner_outputs[CGP_INPUTS + CGP_NODES];

    // copy primary inputs to working array
    memcpy(inner_outputs, inputs, sizeof(cgp_value_t) * CGP_INPUTS);

    for (int i = 0; i < CGP_NODES; i++) {
        _cgp_node *n = &(chr->nodes[i]);
        cgp_value_t A = inner_outputs[n->inputs[0]];
        cgp_value_t B = inner_outputs[n->inputs[1]];
        cgp_value_t Y;

        switch (n->function) {
            case c255:          Y = 255;            break;
            case identity:      Y = A;              break;
            case inversion:     Y = 255 - A;        break;
            case b_or:          Y = A | B;          break;
            case b_not1or2:     Y = ~A | B;         break;
            case b_and:         Y = A & B;          break;
            case b_nand:        Y = ~(A & B);       break;
            case b_xor:         Y = A ^ B;          break;
            case rshift1:       Y = A >> 1;         break;
            case rshift2:       Y = A >> 2;         break;
            case swap:          Y = SWAP(A, B);     break;
            case add:           Y = A + B;          break;
            case add_sat:       Y = ADD_SAT(A, B);  break;
            case avg:           Y = (A + B) >> 1;   break;
            case max:           Y = MAX(A, B);      break;
            case min:           Y = MIN(A, B);      break;
        }

        inner_outputs[CGP_INPUTS + i] = Y;
    }

    for (int i = 0; i < CGP_OUTPUTS; i++) {
        outputs[i] = inner_outputs[chr->outputs[i]];
    }

#ifdef TEST_EVAL
    for (int i = 0; i < CGP_INPUTS + CGP_NODES; i++) {
        printf("%c: %u = %u\n", (i < CGP_INPUTS? 'I' : 'N'), i, inner_outputs[i]);
    }
    for (int i = 0; i < CGP_OUTPUTS; i++) {
        printf("O: %u = %u\n", i, outputs[i]);
    }
#endif
}


/**
 * Calculate fitness of given chromosome, but only if its `has_fitness`
 * attribute is set to `false`
 * @param chr
 */
cgp_fitness_t cgp_evaluate_chr(cgp_chr chr)
{
    if (!chr->has_fitness) {
        return cgp_reevaluate_chr(chr);
    } else {
        return chr->fitness;
    }
}


/**
 * Calculate fitness of given chromosome, regardless of its `has_fitness`
 * value
 * @param chr
 */
cgp_fitness_t cgp_reevaluate_chr(cgp_chr chr)
{
    chr->fitness = (*fitness_function)(chr);
    chr->has_fitness = true;
    return chr->fitness;
}


/**
 * Replace gene on given locus with random alele
 * @param chr
 * @param gene
 */
void cgp_randomize_gene(cgp_chr chr, int gene)
{
    if (gene >= CGP_CHR_LENGTH)
        return;

    if (gene < CGP_CHR_OUTPUTS_INDEX) {
        // mutating node input or function
        int node_index = gene / 3;
        int gene_index = gene % 3;
        int col = cgp_node_col(node_index);

        TEST_RANDOMIZE_PRINTF("gene %u, node %u, gene %u, col %u\n", gene, node_index, gene_index, col);

        if (gene_index == CGP_FUNC_INPUTS) {
            // mutating function
            chr->nodes[node_index].function = rand_range(0, CGP_FUNC_COUNT - 1);
            TEST_RANDOMIZE_PRINTF("func 0 - %u\n", CGP_FUNC_COUNT - 1);

        } else {
            // mutating input
            chr->nodes[node_index].inputs[gene_index] = rand_schoice(allowed_vals[col].size, allowed_vals[col].values);
            TEST_RANDOMIZE_PRINTF("input choice from %u\n", allowed_vals[col].size);
        }

    } else {
        // mutating primary output connection
        int index = gene - CGP_CHR_OUTPUTS_INDEX;
        chr->outputs[index] = rand_range(0, CGP_INPUTS + CGP_NODES - 1);
        TEST_RANDOMIZE_PRINTF("out 0 - %u\n", CGP_INPUTS + CGP_NODES - 1);
    }
}


/* population *****************************************************************/


/**
 * Create a new CGP population with given size
 * @param  size
 * @return
 */
cgp_pop cgp_create_pop(int size)
{
    cgp_pop new_pop = (cgp_pop) malloc(sizeof(_cgp_pop));
    if (new_pop == NULL) return NULL;

    new_pop->size = size;
    new_pop->best_chr_index = -1;
    new_pop->chromosomes = (cgp_chr*) malloc(sizeof(cgp_chr) * size);
    if (new_pop->chromosomes == NULL) {
        free(new_pop);
        return NULL;
    }

    for (int i = 0; i < size; i++) {
        new_pop->chromosomes[i] = cgp_create_chr();
    }

    return new_pop;
}


/**
 * Clear memory associated with given population (including its chromosomes)
 * @param pop
 */
void cgp_destroy_pop(cgp_pop pop)
{
    if (pop != NULL) {
        for (int i = 0; i < pop->size; i++) {
            cgp_destroy_chr(pop->chromosomes[i]);
        }
        free(pop->chromosomes);
    }
    free(pop);
}


/**
 * Calculate fitness of whole population, using `cgp_evaluate_chr`
 * @param chr
 */
void cgp_evaluate_pop(cgp_pop pop)
{
    cgp_fitness_t best_fitness;
    int best_index;

    // reevaluate population
    for (int i = 0; i < pop->size; i++) {
        cgp_fitness_t f = cgp_evaluate_chr(pop->chromosomes[i]);
        if (i == 0 || cgp_is_better_or_same(f, best_fitness)) {
            best_fitness = f;
            best_index = i;
        }
    }

    // if best index hasn't changed, try to find different one with the same fitness
    if (best_index == pop->best_chr_index) {
        for (int i = 0; i < pop->size; i++) {
            cgp_fitness_t f = pop->chromosomes[i]->fitness;
            if (i != best_index && f == best_fitness) {
                best_index = i;
                break;
            }
        }
    }

    // set new best values
    pop->best_fitness = best_fitness;
    pop->best_chr_index = best_index;
}


/**
 * Advance population to next generation
 * @param pop
 * @param mutation_rate
 */
void cgp_next_generation(cgp_pop pop, int mutation_rate)
{
    cgp_chr parent = pop->chromosomes[pop->best_chr_index];

    for (int i = 0; i < pop->size; i++) {
        if (i == pop->best_chr_index) continue;
        cgp_replace_chr(pop->chromosomes[i], parent);
        cgp_mutate_chr(pop->chromosomes[i], mutation_rate);
    }

    cgp_evaluate_pop(pop);
}
