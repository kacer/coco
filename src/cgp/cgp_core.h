/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master's Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include "../ga.h"


#ifdef SYMREG
  #include "../symreg/cgp.h"
#else
  #include "../ifilter/cgp.h"
#endif


typedef struct {
    int inputs;
    int outputs;
    int cols;
    int rows;
    int lback;
    int mutation_rate;
    ga_fitness_func_t fitness_function;
} cgp_settings_t;


static const ga_problem_type_t CGP_PROBLEM_TYPE = maximize;


/**
 * One CGP node (function block)
 */
typedef struct {
    int inputs[CGP_FUNC_INPUTS];
    cgp_func_t function;
    bool is_active;
    bool is_constant;
    cgp_value_t constant_value;
} cgp_node_t;


/**
 * Chromosome
 */
struct cgp_genome {
    cgp_node_t *nodes;
    int *outputs;

    int inputs_count;
    int outputs_count;
    int cols;
    int rows;
};
typedef struct cgp_genome* cgp_genome_t;


/**
 * Initialize CGP internals
 */
void cgp_init(cgp_settings_t *settings);


/**
 * Deinitialize CGP internals
 */
void cgp_deinit();


/**
 * Create a new CGP population with given size
 * @param  mutation rate (in number of genes)
 * @return
 */
ga_pop_t cgp_init_pop(int pop_size);


/**
 * Allocates memory for new CGP genome
 * @return pointer to newly allocated genome
 */
void* cgp_alloc_genome();


/**
 * Initializes CGP genome to random values
 * @param chromosome
 * @return 0 on success, other value on error
 */
int cgp_randomize_genome(ga_chr_t chromosome);


/**
 * Deinitialize CGP genome
 * @param  genome
 * @return
 */
void cgp_free_genome(void *genome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_copy_genome(void *_dst, void *_src);


/**
 * Replace gene on given locus with random alele
 * @param chr
 * @param gene
 * @return whether active node was changed or not (phenotype has changed)
 */
bool cgp_randomize_gene(cgp_genome_t genome, int gene);


/**
 * Mutate given chromosome
 * @param chr
 */
void cgp_mutate_chr(ga_chr_t chromosome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_replace_chr(ga_chr_t chromosome, ga_chr_t replacement);


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 * @return Whether CGP function has changed and evaluation should be restarted
 */
bool cgp_get_output(ga_chr_t chromosome, cgp_value_t *inputs, cgp_value_t *outputs);


/**
 * Calculates CGP node output value.
 *
 * Defined in symreg/cgp.c or ifilter/cgp.c
 *
 * @param  n CGP node
 * @param  A Input value A
 * @param  B Input value B
 * @param  Y Output value
 * @return Whether CGP function changed and evaluation should be restarted
 */
bool cgp_get_node_output(cgp_node_t *n, cgp_value_t A,
    cgp_value_t B, cgp_value_t *Y);


/**
 * Create new generation
 * @param pop
 * @param mutation_rate
 */
void cgp_offspring(ga_pop_t pop);


/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_nodes_count(cgp_genome_t genome)
{
    return genome->rows * genome->rows;
}


/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_output_genes_offset(cgp_genome_t genome)
{
    return ((CGP_FUNC_INPUTS + 1) * cgp_nodes_count(genome));
}


/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_genome_length(cgp_genome_t genome)
{
    return cgp_output_genes_offset(genome) + genome->outputs_count;
}



/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_node_index(cgp_genome_t genome, int col, int row)
{
    return genome->rows * col + row;
}


/**
 * Returns node column based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_col(cgp_genome_t genome, int index)
{
    return index / genome->rows;
}


/**
 * Returns node row based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_row(cgp_genome_t genome, int index)
{
    return index % genome->rows;
}


/**
 * Finds which blocks are active.
 * @param chromosome
 * @param active
 */
void cgp_find_active_blocks(ga_chr_t chromosome);


/**
 * Returns l-back parameter value
 */
int cgp_get_lback();
