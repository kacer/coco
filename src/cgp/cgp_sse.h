/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master's Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include <immintrin.h>

#include "../cgp/cgp_core.h"


static const int SSE2_BLOCK_SIZE = sizeof(__m128i) / sizeof(cgp_value_t);


typedef __m128i __m128i_aligned __attribute__ ((aligned (16)));


/**
 * Calculate output of given chromosome and inputs using SSE instructions
 *
 * The output_data array is in fact two-dimensional and it should be
 * at least (SSE2_BLOCK_SIZE * CGP_OUTPUTS) big.
 *
 * @param chr
 * @param input_data
 * @param input_offset offset from which to copy input data to SSE registers
 * @param output_data will be written here (two-dimensional array!)
 * @return number of items written to output_data
 */
int cgp_sse_get_output(
    ga_chr_t chromosome,
    cgp_value_t *input_data[CGP_INPUTS],
    int input_offset,
    cgp_value_t *output_data,
    bool *should_restart);


/**
 * Calculates CGP node output value using SSE instructions
 *
 * @param  n CGP node
 * @param  A Input value A
 * @param  B Input value B
 * @param Whether CGP function has changed and evaluation should be restarted
 * @return Output value
 */
__m128i cgp_sse_get_node_output(
    cgp_node_t *n,
    __m128i A,
    __m128i B,
    bool *should_restart);


/**
 * Extracts output value from returned block
 * @param  output_data
 * @param  output_idx
 * @param  offset
 * @return
 */
static inline cgp_value_t cgp_sse_extract_output(
    cgp_value_t *output_data,
    int output_idx,
    int offset)
{
    return output_data[output_idx * SSE2_BLOCK_SIZE + offset];
}
