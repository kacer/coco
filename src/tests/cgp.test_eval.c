/**
 * Tests CGP evaluation = calculation of the outputs.
 * Compile with -DTEST_EVAL
 */

#include <stdlib.h>
#include <stdio.h>

#include "../cgp.h"


int main(int argc, char const *argv[])
{
    cgp_value_t inputs[CGP_INPUTS] = {0, 1, 2, 3, 4, 5, 6, 7, 8};
    cgp_value_t outputs[CGP_OUTPUTS] = {};

    cgp_init();

    cgp_genome_t genome = (cgp_genome_t) malloc(sizeof(struct cgp_genome));
    struct ga_chr chr = {
        .genome = genome
    };

    // define nodes
    for (int i = 0; i < CGP_NODES; i++) {
        cgp_node_t *n = &(genome->nodes[i]);

        n->inputs[0] = i + CGP_INPUTS - CGP_ROWS - 3;
        n->inputs[1] = i + CGP_INPUTS - CGP_ROWS - 1;
        n->function = (cgp_func_t) (i % CGP_FUNC_COUNT);
    }

    // define outputs
    genome->outputs[0] = 13;

    cgp_dump_chr_asciiart(&chr, stdout);
    putchar('\n');
    cgp_get_output(&chr, inputs, outputs);

    free(genome);
    cgp_deinit();
}
