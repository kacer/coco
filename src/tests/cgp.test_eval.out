Inputs: 9
Outputs: 1
Size: 8 x 4
Blocks: 2-ary, 1 output(s), 16 functions
Fitness: <none>
     .------------------------------------------------------------------------------------------------------------------------------------------------.
     |      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
[ 0]>| [ 2]>|    |>[ 9]  [ 6]>|    |>[13]  [10]>|    |>[17]  [14]>|    |>[21]  [18]>|    |>[25]  [22]>|    |>[29]  [26]>|    |>[33]  [30]>|    |>[37] |>[13]
[ 1]>| [ 4]>| FF |       [ 8]>|~1|2|       [12]>|a>>1|       [16]>| +S |       [20]>| FF |       [24]>|~1|2|       [28]>|a>>1|       [32]>| +S |      |
[ 2]>|      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
[ 3]>|      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
[ 4]>| [ 3]>|    |>[10]  [ 7]>|    |>[14]  [11]>|    |>[18]  [15]>|    |>[22]  [19]>|    |>[26]  [23]>|    |>[30]  [27]>|    |>[34]  [31]>|    |>[38] |
[ 5]>| [ 5]>|  a |       [ 9]>| and|       [13]>|a>>2|       [17]>| avg|       [21]>|  a |       [25]>| and|       [29]>|a>>2|       [33]>| avg|      |
[ 6]>|      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
[ 7]>|      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
[ 8]>| [ 4]>|    |>[11]  [ 8]>|    |>[15]  [12]>|    |>[19]  [16]>|    |>[23]  [20]>|    |>[27]  [24]>|    |>[31]  [28]>|    |>[35]  [32]>|    |>[39] |
     | [ 6]>|FF-a|       [10]>|nand|       [14]>|swap|       [18]>| max|       [22]>|FF-a|       [26]>|nand|       [30]>|swap|       [34]>| max|      |
     |      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
     |      .----.            .----.            .----.            .----.            .----.            .----.            .----.            .----.      |
     | [ 5]>|    |>[12]  [ 9]>|    |>[16]  [13]>|    |>[20]  [17]>|    |>[24]  [21]>|    |>[28]  [25]>|    |>[32]  [29]>|    |>[36]  [33]>|    |>[40] |
     | [ 7]>| or |       [11]>| xor|       [15]>| +  |       [19]>| min|       [23]>| or |       [27]>| xor|       [31]>| +  |       [35]>| min|      |
     |      '----'            '----'            '----'            '----'            '----'            '----'            '----'            '----'      |
     '------------------------------------------------------------------------------------------------------------------------------------------------'

I: 0 = 0
I: 1 = 1
I: 2 = 2
I: 3 = 3
I: 4 = 4
I: 5 = 5
I: 6 = 6
I: 7 = 7
I: 8 = 8
N: 9 = 255
N: 10 = 7
N: 11 = 249
N: 12 = 13
N: 13 = 255
N: 14 = 1
N: 15 = 254
N: 16 = 242
N: 17 = 121
N: 18 = 63
N: 19 = 30
N: 20 = 241
N: 21 = 255
N: 22 = 46
N: 23 = 63
N: 24 = 121
N: 25 = 255
N: 26 = 63
N: 27 = 209
N: 28 = 255
N: 29 = 255
N: 30 = 17
N: 31 = 238
N: 32 = 0
N: 33 = 0
N: 34 = 59
N: 35 = 30
N: 36 = 255
N: 37 = 255
N: 38 = 44
N: 39 = 59
N: 40 = 0
O: 0 = 255
